src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -569,6 +574,62 @@ private bool DelegateNeedsReturn(MethodSymbol invokeMethod)
             return true;
         }
 
+        private BoundLambda ReallyBindNaturalType(TypeSymbol expressionType)
+        {
+            var compilation = Binder.Compilation;
+            var parameterRefKindsBuilder = ArrayBuilder<RefKind>.GetInstance(ParameterCount);
+            var parameterTypesBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance(ParameterCount);
+            for (int i = 0; i < ParameterCount; i++)
+            {
+                parameterRefKindsBuilder.Add(HasExplicitlyTypedParameterList ? RefKind(i) : default);
+                parameterTypesBuilder.Add(HasExplicitlyTypedParameterList ? ParameterTypeWithAnnotations(i) : TypeWithAnnotations.Create(compilation.GetSpecialType(SpecialType.System_Object)));
+            }
+            var parameterRefKinds = parameterRefKindsBuilder.ToImmutableAndFree();
+            var parameterTypes = parameterTypesBuilder.ToImmutableAndFree();
+
+            var lambdaSymbol = new LambdaSymbol(
+                Binder,
+                compilation,
+                Binder.ContainingMemberOrLambda!,
```

Comment:
How do we know?

Reply:
We assert `ContainingMemberOrLambda != null` in the `UnboundLambdaState` constructor, line 439.

Reply:
I don't want to block on this, but our guideline is to use assertion unless check (direct or indirect) already exists in method. Recently you've been also extending that to dereferences (would blow up anyways), which I'm okay with.
But this is neither

---
In reply to: [616975571](https://github.com/dotnet/roslyn/pull/52448#discussion_r616975571) [](ancestors = 616975571)

Reply:
> our guideline is to use assertion unless check (direct or indirect) already exists in method

Added asserts.